cli/client.go

```diff
@@ -1732,13 +1757,188 @@ func (conf *Config) checkUpdate(c *cli.Context) error {
+	if err != nil {
+		return "", errors.Errorf("binary download failed: %v", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		utils.UncheckedError(resp.Body.Close())
+		return "", errors.Errorf("binary download failed: server returned status %d", resp.StatusCode)
 	}
 
+	// create a temp file that we write the downloaded binary into
+	goos := runtime.GOOS
+	tempFileName := "viam-cli-update"
+	if goos == osWindows {
+		tempFileName += ".exe"
+	}
+	tempFileName += ".new"
+
+	// Create the temp file in the same directory as the binary
+	latestBinaryPath := filepath.Join(directoryPath, tempFileName)
+	latestBinaryFile, err := os.Create(latestBinaryPath) //nolint:gosec
+	if err != nil {
+		utils.UncheckedError(resp.Body.Close())
+		if os.IsPermission(err) {
+			if goos == osWindows {
+				return "", errors.New("permission denied: run PowerShell as Administrator")
+			}
+			return "", errors.New("permission denied: run 'sudo viam self-update'")
+		}
+		return "", errors.Errorf("failed to create temp file: %v", err)
+	}
+
+	// Write downloaded content to temp file
+	_, err = io.Copy(latestBinaryFile, resp.Body)
+	utils.UncheckedError(resp.Body.Close())
+	utils.UncheckedError(latestBinaryFile.Close())
+	if err != nil {
+		return "", errors.Errorf("failed to write downloaded binary: %v", err)
+	}
+
+	// Make executable on Unix-like systems, if permissions are improperly set then no
+	// change will be made and user has to run sudo
+	if goos != osWindows {
+		if err := os.Chmod(latestBinaryPath, 0o755); err != nil { //nolint:gosec
+			if os.IsPermission(err) {
+				return "", errors.New("permission denied: run 'sudo viam self-update'")
+			}
+			return "", errors.Errorf("failed to make binary executable: %v", err)
+		}
+	}
+	return latestBinaryPath, nil
+}
+
+func replaceBinary(localBinaryPath, latestBinaryPath string) error {
+	// Windows doesn't allow renaming running executables, so we rename the old one
+	// and put the new one in the original location to be used next run
+	if runtime.GOOS == osWindows {
+		// make sure no leftover binary files with .old
+		oldPath := localBinaryPath + ".old"
+		os.Remove(oldPath) //nolint:errcheck, gosec
+		// 1. rename old binary to .old
+		if err := os.Rename(localBinaryPath, oldPath); err != nil {
+			if os.IsPermission(err) {
+				return errors.New("permission denied: run PowerShell as Administrator")
+			}
+			return errors.Errorf("failed to rename old binary: %v", err)
+		}
+
+		// 2. rename new binary to original path
+		if err := os.Rename(latestBinaryPath, localBinaryPath); err != nil {
+			if os.IsPermission(err) {
+				return errors.New("permission denied: run PowerShell as Administrator")
+			}
+			// If this fails, try to restore old binary by renaming back (removing .old)
+			if restoreErr := os.Rename(oldPath, localBinaryPath); restoreErr != nil {
+				return errors.Errorf("failure in self-update: please redownload the CLI binary at https://docs.viam.com/cli/#install")
```

Comment:
(minor) probably good to be consistent with usage of `update` vs. `self-update`.